pkg/image/prune/prune.go

```diff
@@ -685,57 +701,76 @@ func pruneStreams(g graph.Graph, imageNodes []*imagegraph.ImageNode, streamPrune
 				continue
 			}
 
-			if err := retry.RetryOnConflict(retry.DefaultRetry, func() error {
-				stream, err := streamPruner.GetImageStream(streamNode.ImageStream)
-				if err != nil {
-					return err
-				}
-				updatedTags := sets.NewString()
+			streamName := getName(streamNode.ImageStream)
+
+			if _, ok := prunableStreams[streamName]; !ok {
+				prunableStreams[streamName] = streamNode
+			}
+		}
+	}
+
+	for streamName, streamNode := range prunableStreams {
+		glog.V(4).Infof("Checking if ImageStream %s", streamName)
 
-				glog.V(4).Infof("Checking if ImageStream %s has references to image %s in status.tags", getName(stream), imageNode.Image.Name)
+		err := retry.RetryOnConflict(retry.DefaultRetry, func() error {
+			stream, err := streamPruner.GetImageStream(streamNode.ImageStream)
+			if err != nil {
+				return err
+			}
 
-				for tag, history := range stream.Status.Tags {
-					glog.V(4).Infof("Checking tag %q", tag)
+			updatedTags := sets.NewString()
+			deletedTags := sets.NewString()
 
-					newHistory := imageapi.TagEventList{}
+			for tag, history := range stream.Status.Tags {
+				glog.V(4).Infof("Checking tag %q", tag)
 
-					for i, tagEvent := range history.Items {
-						glog.V(4).Infof("Checking tag event %d with image %q", i, tagEvent.Image)
+				newHistory := imageapi.TagEventList{}
 
-						if tagEvent.Image != imageNode.Image.Name {
-							glog.V(4).Infof("Tag event doesn't match deleted image - keeping")
-							newHistory.Items = append(newHistory.Items, tagEvent)
-						} else {
-							glog.V(4).Infof("Tag event matches deleted image - removing reference")
-							updatedTags.Insert(tag)
-						}
-					}
-					if len(newHistory.Items) == 0 {
-						glog.V(4).Infof("Removing tag %q from status.tags of ImageStream %s", tag, getName(stream))
-						delete(stream.Status.Tags, tag)
+				for i, tagEvent := range history.Items {
+					glog.V(4).Infof("Checking tag event %d with image %q", i, tagEvent.Image)
+
+					if _, ok := imageNodes[tagEvent.Image]; !ok {
+						glog.V(4).Infof("Tag event doesn't match deleted image - keeping")
```

Comment:
New code will produce MUCH better logs, but I still don't like them :-) They are not greppable.

```
glog.V(4).Infof("Checking image stream tag %s:%s: event %d with image %q - keeping, doesn't match deleted image", streamName, tag, i, tagEvent.Image)
glog.V(4).Infof("Checking image stream tag %s:%s: event %d with image %q - keeping, within the allowed range of revisions", streamName, tag, i, tagEvent.Image)
glog.V(4).Infof("Checking image stream tag %s:%s: event %d with image %q - removing, matches deleted image", streamName, tag, i, tagEvent.Image)
```